benchmarks/src/bin/external_aggr.rs

```diff
@@ -0,0 +1,390 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! external_aggr binary entrypoint
+
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::OnceLock;
+use structopt::StructOpt;
+
+use arrow::record_batch::RecordBatch;
+use arrow::util::pretty;
+use datafusion::datasource::file_format::parquet::ParquetFormat;
+use datafusion::datasource::listing::{
+    ListingOptions, ListingTable, ListingTableConfig, ListingTableUrl,
+};
+use datafusion::datasource::{MemTable, TableProvider};
+use datafusion::error::Result;
+use datafusion::execution::memory_pool::FairSpillPool;
+use datafusion::execution::memory_pool::{human_readable_size, units};
+use datafusion::execution::runtime_env::RuntimeConfig;
+use datafusion::physical_plan::display::DisplayableExecutionPlan;
+use datafusion::physical_plan::{collect, displayable};
+use datafusion::prelude::*;
+use datafusion_benchmarks::util::{BenchmarkRun, CommonOpt};
+use datafusion_common::instant::Instant;
+use datafusion_common::{exec_datafusion_err, exec_err, DEFAULT_PARQUET_EXTENSION};
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "datafusion-external-aggregation",
+    about = "DataFusion external aggregation benchmark"
+)]
+enum ExternalAggrOpt {
+    Benchmark(ExternalAggrConfig),
+}
+
+#[derive(Debug, StructOpt)]
+struct ExternalAggrConfig {
+    /// Query number. If not specified, runs all queries
+    #[structopt(short, long)]
+    query: Option<usize>,
+
+    /// Memory limit (e.g. '100M', '1.5G'). If not specified, run all pre-defined memory limits for given query.
+    #[structopt(long)]
+    memory_limit: Option<String>,
+
+    /// Common options
+    #[structopt(flatten)]
+    common: CommonOpt,
+
+    /// Path to data files (lineitem). Only parquet format is supported
+    #[structopt(parse(from_os_str), required = true, short = "p", long = "path")]
+    path: PathBuf,
+
+    /// Load the data into a MemTable before executing the query
+    #[structopt(short = "m", long = "mem-table")]
+    mem_table: bool,
+
+    /// Path to JSON benchmark result to be compare using `compare.py`
+    #[structopt(parse(from_os_str), short = "o", long = "output")]
+    output_path: Option<PathBuf>,
+}
+
+struct QueryResult {
+    elapsed: std::time::Duration,
+    row_count: usize,
+}
+
+/// Query Memory Limits
+/// Map query id to predefined memory limits
+///
+/// Q1 requires 36MiB for aggregation
+/// Memory limits to run: 64MiB, 32MiB, 16MiB
+/// Q2 requires 250MiB for aggregation
+/// Memory limits to run: 512MiB, 256MiB, 128MiB, 64MiB, 32MiB
+static QUERY_MEMORY_LIMITS: OnceLock<HashMap<usize, Vec<u64>>> = OnceLock::new();
+
+impl ExternalAggrConfig {
+    const AGGR_TABLES: [&'static str; 1] = ["lineitem"];
+    const AGGR_QUERIES: [&'static str; 2] = [
+        // Q1: Output size is ~25% of lineitem table
+        r#"
+        SELECT count(*)
```

Comment:
Another type of query that is oten run is the classic "top 10"  type that shows up in clickbench a lot:

Something like

```sql
> select l_orderkey, avg(l_extendedprice) as a, avg(l_discount) as d from 'lineitem' GROUP BY l_orderkey ORDER BY a DESC limit 10;
+------------+---------------+----------+
| l_orderkey | a             | d        |
+------------+---------------+----------+
| 3811460    | 104899.500000 | 0.050000 |
| 1744195    | 104649.500000 | 0.090000 |
| 5151266    | 104449.500000 | 0.000000 |
| 4571042    | 104399.500000 | 0.090000 |
| 1198304    | 104299.500000 | 0.020000 |
| 1134944    | 104249.000000 | 0.020000 |
| 2582850    | 104199.000000 | 0.000000 |
| 282754     | 103949.500000 | 0.020000 |
| 2038305    | 103949.000000 | 0.030000 |
| 944835     | 103899.500000 | 0.000000 |
+------------+---------------+----------+
10 row(s) fetched.
Elapsed 0.074 seconds.
```

Reply:
This is an interesting case, sort can also be memory-limited

Reply:
> This is an interesting case, sort can also be memory-limited

It is true, though most engines (including DataFusion) will use `TopK` instead (aka not sort the entire stream, just keep the top 10 values) so this query would likely only spill in aggregation not in sort